### PR TITLE
Static HTML Template helper

### LINKED ISSUE #14529: Build size: innerHTML based element creation
Re: https://github.com/ampproject/amphtml/issues/13906

We have `document.createElement`, `el.setAttribute`, etc calls to generate static DOM trees all over our codebase. For truly static content, we should be providing a `html` helper to generate these trees:

```js
function html(strings) {
  // We don't handle expressions, only fully static trees.
  if (strings.length !== 1) throw new Error('non static template detected');

  const div = document.createElement('div');
  div.innerHTML = strings[0];
  return div.firstElementChild;
}

class Example extends BaseElement {
  buildCallback() {
    const div = html`<div attr="whatever"><p>some static tree</p></div>`;

    console.assert(div.tagName === 'div');
  }
}
```

/cc @alanorozco

### PR DESCRIPTION
**Pull request by jridgewell**
Create static HTML DOM trees using a HTML like syntax. Leading/trailing text (including whitespace) is dropped, and only a single root element in the template is supported.

This is **only** available for static trees. **Dynamic variables in the string are not permitted**, and will result in linting errors and runtime errors.

Valid uses are:

```js
// Yes, the variable name must be "html".
const html = htmlFor(doc || el);

const div = html`<div><p>Hello World</p></div>`;
```

Or, you can use it directly as the template literal tag:

```js
const div = htmlFor(doc || el)`<div><p>Hello World</p></div>`;
```

Closes #14529, jump starts #13906.
/cc @alanorozco, @kristoferbaxter

### CODE REVIEW

File: src/service/viewport/viewport-binding-ios-embed-wrapper.js [hunk 1]
```diff
@@ -44,14 +45,15 @@ export class ViewportBindingIosEmbedWrapper_ {
     /** @const {!Window} */
     this.win = win;
 
-    const topClasses = this.win.document.documentElement.className;
-    this.win.document.documentElement.className = '';
-    this.win.document.documentElement.classList.add('i-amphtml-ios-embed');
+    const documentElement = this.win.document.documentElement;
+    const topClasses = documentElement.className;
+    documentElement.className = 'i-amphtml-ios-embed';
```
Comment (kristoferbaxter) [inline src/service/viewport/viewport-binding-ios-embed-wrapper.js]: Nice clean up. 

I'm curious why className was set to an empty value and the classList was added to afterward in the previous code. Is there a subtle behaviour specific to iOS?

File: src/layout.js [hunk 2]
```diff
@@ -444,14 +445,15 @@ export function applyStaticLayout(element) {
...
-    const sizer = element.ownerDocument.createElement('i-amphtml-sizer');
-    const intrinsicSizer = element.ownerDocument.createElement('img');
-    sizer.classList.add('i-amphtml-sizer');
-    intrinsicSizer.classList.add('i-amphtml-intrinsic-sizer');
+    const sizer = htmlFor(element)`
+      <i-amphtml-sizer class="i-amphtml-sizer">
+        <img class="i-amphtml-intrinsic-sizer" />
+      </i-amphtml-sizer>`;
+    const intrinsicSizer = sizer.firstElementChild;
     intrinsicSizer.setAttribute('src',
```
Comment (kristoferbaxter) [inline src/layout.js]: Perhaps I am missing something. Why can't this attribute be included in the template literal?

File: src/preconnect.js [hunk 3]
```diff
@@ -226,10 +227,9 @@ class PreconnectService {
   }
 
   performPreload_(url) {
-    const preload = this.document_.createElement('link');
-    preload.setAttribute('rel', 'preload');
+    const preload = htmlFor(this.document_)`
+        <link rel="preload" referrerpolicy="origin" />`;
     preload.setAttribute('href', url);
```
Comment (kristoferbaxter) [inline src/preconnect.js]: Same question here, is there a reason the url cannot be included here in the tagged template literal?

Comment (jridgewell) [reply src/preconnect.js]: The url is not statically analyzable, so there's not guarantee that it's not malicious.

File: src/service/viewport/viewport-binding-ios-embed-wrapper.js (hunk 1 above)
Comment (jridgewell) [reply src/service/viewport/viewport-binding-ios-embed-wrapper.js]: Looks like an oversight from https://github.com/ampproject/amphtml/pull/6276/files#diff-6eedd5ca45df1b5aee7d616b017d360aR1419.

**Comment by alanorozco** [comment]
/cc @newmuis @Enriqe @gmajoulet for `amp-story`.

IMO this should be used instead of `SimpleTemplate` where the templates are static.

Dynamic templating still unsolved.

**Comment by newmuis** [comment]
@alanorozco Thanks to our naive localization approach, pretty much none of our templates are static anymore (translated strings get swapped in at runtime) ☹️

**Comment by jridgewell** [comment]
>  (translated strings get swapped in at runtime)

Using the `html` helper is still a better approach here:

```diff
diff --git a/extensions/amp-story/0.1/amp-story.js b/extensions/amp-story/0.1/amp-story.js
--- a/extensions/amp-story/0.1/amp-story.js
+++ b/extensions/amp-story/0.1/amp-story.js
  * Container for "pill-style" share widget, rendered on desktop.
  * @private @const {!./simple-template.ElementDef}
  */
-const SHARE_WIDGET_PILL_CONTAINER = {
-  tag: 'div',
-  attrs: dict({'class': 'i-amphtml-story-share-pill'}),
-  children: [
-    {
-      tag: 'span',
-      attrs: dict({'class': 'i-amphtml-story-share-pill-label'}),
-      localizedStringId:
-          LocalizedStringId.AMP_STORY_SYSTEM_LAYER_SHARE_WIDGET_LABEL,
-    },
-  ],
-};
+const shareWidgetPillContainer = html`
+    <div class="i-amphtml-story-share-pill">
+      <span class="i-amphtml-story-share-pill-label"></span>
+    </div>`
+shareWidgetPillContainer.querySelector('i-amphtml-story-share-pill-label')
+    .innerText = LocalizedStringId.AMP_STORY_SYSTEM_LAYER_SHARE_WIDGET_LABEL;
```

**Comment by jridgewell** [comment]
| | Build | Dist |
| --- | ---- | ---- |
| Pre: | 334.31KB | 76.6KB |
| Post: | 335.54KB | 76.84KB |
| After #14657 and #14658 | | 76.72KB |

**Pull request merged**